### PR TITLE
ensuring connection timeouts get wrapped by MavenDownloadingException

### LINKED ISSUE #4080: MavenPomDownloader - UncheckedIOException does not get wrapped into MavenDownloadingException
## What version of OpenRewrite are you using?

I am using

- rewrite @ main

## How are you running OpenRewrite?

Unit tests

## What is the smallest, simplest way to reproduce the problem?

In my walled-garden network environment, several tests fail from `MavenPomDownloaderTest`, such as `skipsLocalInvalidArtifactsEmptyJar`, with stack traces like this:

```
org.opentest4j.AssertionFailedError: Unexpected exception type thrown, 
Expected :class org.openrewrite.maven.MavenDownloadingException
Actual   :class java.lang.RuntimeException
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3115)
	at org.openrewrite.maven.internal.MavenPomDownloaderTest.skipsLocalInvalidArtifactsEmptyJar(MavenPomDownloaderTest.java:729)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: java.net.SocketTimeoutException: Connect timed out
	at org.openrewrite.maven.internal.MavenPomDownloader.requestAsAuthenticatedOrAnonymous(MavenPomDownloader.java:810)
	at org.openrewrite.maven.internal.MavenPomDownloader.download(MavenPomDownloader.java:572)
	at org.openrewrite.maven.internal.MavenPomDownloaderTest.lambda$skipsLocalInvalidArtifactsEmptyJar$13(MavenPomDownloaderTest.java:731)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
	... 6 more
Caused by: java.io.UncheckedIOException: java.net.SocketTimeoutException: Connect timed out
	at org.openrewrite.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:116)
	at org.openrewrite.maven.internal.MavenPomDownloader.lambda$sendRequest$1(MavenPomDownloader.java:142)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at org.openrewrite.maven.internal.MavenPomDownloader.sendRequest(MavenPomDownloader.java:141)
	at org.openrewrite.maven.internal.MavenPomDownloader.requestAsAuthenticatedOrAnonymous(MavenPomDownloader.java:802)
	... 9 more
Caused by: java.net.SocketTimeoutException: Connect timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:551)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:602)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:178)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:534)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:639)
	at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266)
	at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:380)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:193)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1242)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1128)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:179)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1665)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:308)
	at org.openrewrite.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:97)
	... 18 more

```

## What did you expect to see?

Happy test, or, perhaps an easy-to-read failure from being unable to connect to the default remote Maven repo in my environment.

## What did you see instead?

The resultant exception is not wrapped into `MavenDownloadingException` and therefore does not give useful detail about the exact request/artifact that led to an error.

So when this happens for real recipe runs, that lack of detail is really inconvenient.

## What is the full stack trace of any errors you encountered?

Seen above

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

Yes, but, it's a little tricky because I know I'm in a non-standard network. I'd ideally like to define a unit test which simulates the "connect timeout" behavior, but, hard for me to know for sure that any given test will achieve that same behavior both in my network and in a "standard" network. I might have to spam Actions a little more than usual to validate that.

### PR DESCRIPTION
**Pull request by nmck257**
Closes #4080.

Might take a couple iterations in Actions; I wrote/tested this in a weird network environment. Fingers crossed.

### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [X] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java [hunk 1]
```diff
@@ -769,8 +771,6 @@ public MavenRepository normalizeRepository(MavenRepository originalRepository, M
                                             repository.getSnapshots(),
                                             repository.getUsername(),
                                             repository.getPassword());
-                                } else if (!e.isClientSideException()) {
-                                    return originalRepository;
```
Comment (nmck257) [inline rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java:773]: removing this feels a little cryptic, but I'm pretty sure it's no longer relevant?
eg, when it was written, the `e.isServerReached` call a few lines earlier did not exist:
https://github.com/openrewrite/rewrite/commit/3bcafba698a442818b740140ce42571a2a54dd7d#diff-a39c7415cc78264a78509aaaba13d76b4eeea28cdc5723e15615b7a8add5b325

Comment (nmck257) [reply rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java:773]: that said, tests pass -- ready for review @timtebeek

File: rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java [hunk 2]
```diff
@@ -149,6 +149,8 @@ byte[] sendRequest(HttpSender.Request request) throws Throwable {
             });
         } catch (FailsafeException failsafeException) {
             throw failsafeException.getCause() == null ? failsafeException : failsafeException.getCause();
+        } catch (Throwable e) {
+            throw new HttpSenderResponseException(e, null, "");
```
Comment (timtebeek) [inline rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java:153]: I worry a bit about how broadly we match any Throwable here, and then throw a `HttpSenderResponseException` despite the underlying issue having been an `UncheckedIOException`, with no guarantee a server response was involved here (as in your case the connection timed out). Elsewhere in `MavenPomDownloader` we use `HttpSenderResponseException` to signify that the server exists, with implications for caching that information beyond just the request here.

https://github.com/openrewrite/rewrite/blob/b67be0403ee9e5eea0e718642bda0ee7f0e8bbb6/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java#L262-L267

Perhaps we should catch and unwrap the UncheckedIOException instead?

```suggestion
        } catch (UncheckedIOException e) {
            throw e.getCause();
```


That's the most likely case of unhandled exceptions looking at how we send requests

https://github.com/openrewrite/rewrite/blob/03c458404073082275a752063887c188d213afbe/rewrite-core/src/main/java/org/openrewrite/ipc/http/HttpUrlConnectionSender.java#L115-L117

Any thoughts on that?

Comment (nmck257) [reply rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java:153]: that does seem better -- will push some changes tomorrow

Comment (timtebeek) [reply rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java]: Thanks! I've gone ahead and applied my suggestion, such that we can go into tomorrow knowing what effect this has on the tests already.

**Comment by timtebeek** [comment]
Two new test failures with the new approach: https://ge.openrewrite.org/s/p33d2lkyakx4u/tests/overview?outcome=FAILED

`listenerRecordsFailedRepositoryAccess()`
```
java.lang.AssertionError: 	
Expecting throwable message:	
  "65b2442b-44ab-4f5c-a7c4-ec45d42f42a3.com"	
to contain:	
  "java.net.UnknownHostException"	
but did not.	
Throwable that failed the check:	
java.net.UnknownHostException: 65b2442b-44ab-4f5c-a7c4-ec45d42f42a3.com	
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:572)	
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
```

`connectTimeout()`
```
Expecting actual throwable to be an instance of:	
  org.openrewrite.maven.MavenDownloadingException	
but was:	
  java.lang.RuntimeException: java.net.SocketTimeoutException: Connect timed out	
	at org.openrewrite.maven.internal.MavenPomDownloader.requestAsAuthenticatedOrAnonymous(MavenPomDownloader.java:810)	
	at org.openrewrite.maven.internal.MavenPomDownloader.download(MavenPomDownloader.java:574)	
	at org.openrewrite.maven.internal.MavenPomDownloaderTest.lambda$connectTimeout$14(MavenPomDownloaderTest.java:752)	
	...(106 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
```

**Comment by nmck257** [comment]
pushed further changes, focused on 1) fixing those test failures and 2) using specific checked exception types instead of `Throwable` in method signatures

**Review by timtebeek — APPROVED** [review]
Thanks a lot! Moving away from `throwable` and using a checked `IOException` does really help take some of the magic out on what can and can't happen. Good to have this case caught & handled correctly, thanks!

**Pull request merged**